### PR TITLE
ARROW-13187: [Python] Avoid creating reference cycle when reading CSV file

### PR DESCRIPTION
**Pull request by pitrou**
Some Python versions have a bug where `signal.getsignal` creates a reference cycle holding execution frames alive (https://bugs.python.org/issue42248).

This would cause excessive lifetimes of the PyArrow table returned by `read_csv`.

### CODE REVIEW

**Comment by pitrou** [comment]
cc @westonpace

File: python/pyarrow/error.pxi [hunk 1]
```diff
@@ -24,6 +24,9 @@ from contextlib import contextmanager
 import os
 import signal
 import threading
+import weakref
```
Comment (westonpace) [inline python/pyarrow/error.pxi:27]: Do you need weakref here?

**Review by westonpace — APPROVED** [review]
Thanks for putting this fix in! LGTM.

**Review by jorisvandenbossche — APPROVED** [review]
I won't say I fully understand this code ;), but LGTM

File: python/pyarrow/error.pxi (hunk 1 above)
Comment (pitrou) [reply python/pyarrow/error.pxi:27]: No, it's just a remnant from a failed attempt.

**Pull request closed without merging**